### PR TITLE
Inlined boxes

### PR DESCRIPTION
**Pull request by lukemaurer**
When generating cmm for `Box_number`, inline the variable if it's used exactly once. This fixes some cases of allocations that are used in only one branch of a switch getting allocated outside of the switch.

### CODE REVIEW

**Comment by lthls** [comment]
I'm mostly fine with the approach. However, I don't like how function calls are handled. The aim of the new logic is to ensure that boxing operations don't get moved past a function call; for function calls whose return continuation is not inlined, there should be nothing to do: all bindings are flushed before the call anyway. For the other cases, rather than the current approach of flushing the box allocations before the call, I'd prefer if `let_bind_expr` took an extra flag (`allocation_barrier`, or `flush_allocations`) that would, if set, convert all box bindings to regular bindings (with the appropriate effects) before adding the new binding. This flag would then be set only on the two relevant cases.

**Comment by lukemaurer** [comment]
> all bindings are flushed before the call anyway.

Actually, pure bindings that are going to be inlined are normally kept around by `flush_delayed_lets` (I think this is fairly new behavior). So we do need some new logic there.

> I'd prefer if let_bind_expr took an extra flag (allocation_barrier, or flush_allocations) that would, if set, convert all box bindings to regular bindings (with the appropriate effects) before adding the new binding.

I think this would be hard to follow: dealing with inlined boxes and adding a let binding are separate concerns. Converting the boxes to effectful computations (by putting them in `env.stages`, I assume?) could work; in any case I'd rather make it a separate operation from `let_expr_bind`.

**Comment by lthls** [comment]
> > all bindings are flushed before the call anyway.
> 
> Actually, pure bindings that are going to be inlined are normally kept around by `flush_delayed_lets` (I think this is fairly new behavior). So we do need some new logic there.

I don't think so. The remaining variables (pure or boxed allocations) are not used, otherwise they would have already been placed; so we can safely ignore them.

> > I'd prefer if let_bind_expr took an extra flag (allocation_barrier, or flush_allocations) that would, if set, convert all box bindings to regular bindings (with the appropriate effects) before adding the new binding.
> 
> I think this would be hard to follow: dealing with inlined boxes and adding a let binding are separate concerns. Converting the boxes to effectful computations (by putting them in `env.stages`, I assume?) could work; in any case I'd rather make it a separate operation from `let_expr_bind`.

To me, `let_expr_bind` is the function that is responsible for preserving the evaluation order. If there are constraints on the order between function calls and boxed allocations, then it seems much better to reflect the constraint at this point.

Anyway, I think this discussion is not very productive without some concrete suggestions; I'll make a PR to your branch so that you can get a better idea of what I think the code should look like.

File: middle_end/flambda2/to_cmm/to_cmm_env.ml [hunk 1]
```diff
@@ -360,6 +362,16 @@ let classify effs =
...
   | No_effects, No_coeffects -> Pure
 
+let is_inlinable_box effs ~extra =
+  (* [effs] is the effects and coeffects of some primitive operation, arising
+     either from the primitive itself or its arguments. If this is a boxing
+     operation (as indicated by [extra]), then we want to inline the box, but
+     this involves moving the arguments, so they must be pure (or at most
+     generative). *)
+  match (effs : Effects_and_coeffects.t), (extra : extra_info option) with
+  | (Only_generative_effects _, No_coeffects), Some Box -> true
```
Comment (lthls) [inline middle_end/flambda2/to_cmm/to_cmm_env.ml:372]: I believe that we won't get `Pure` boxed allocations, as the boxing itself will always contain generative effects, but it would read better (and match the comment above) if we allowed `Pure` effects too.

**Review by lthls — APPROVED** [review]
Looks good to me.
I've left a comment that you could address if you want, but I wouldn't consider it necessary.

File: middle_end/flambda2/to_cmm/to_cmm_env.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml]: Fixed

**Review by mshinwell — APPROVED** [review]
